lib/utils/findMediaFeatureNamesSearch.js

```diff
@@ -0,0 +1,147 @@
+'use strict';
+
+const { TokenType, isToken, stringify, tokenize } = require('@csstools/css-tokenizer');
+const {
+	isTokenNode,
+	parseCommaSeparatedListOfComponentValues,
+	isSimpleBlockNode,
+} = require('@csstools/css-parser-algorithms');
+const {
+	isGeneralEnclosed,
+	isMediaFeature,
+	isMediaQueryInvalid,
+	parseFromTokens,
+} = require('@csstools/media-query-list-parser');
+
+/** @typedef {Array<import('@csstools/media-query-list-parser').MediaQuery>} MediaQueryList */
+/** @typedef {import('@csstools/css-tokenizer').TokenIdent} TokenIdent */
+/** @typedef {{ stringify: () => string }} MediaQuerySerializer */
+
+const rangeFeatureOperator = /[<>=]/;
+
+/**
+ * Search a CSS string for Media Feature names.
+ * For every found name, invoke the callback, passing the token
+ * as an argument.
+ *
+ * Found tokens are mutable and modifications made to them will be reflected in the output.
+ *
+ * This function supports some non-standard syntaxes like SCSS variables and interpolation.
+ *
+ * @param {string} mediaQueryParams
+ * @param {(mediaFeatureName: TokenIdent) => void} callback
+ *
+ * @returns {MediaQuerySerializer}
+ */
+module.exports = function findMediaFeatureNamesSearch(mediaQueryParams, callback) {
+	const tokens = tokenize({ css: mediaQueryParams });
+	const list = parseCommaSeparatedListOfComponentValues(tokens);
+
+	const mediaQueryConditions = list.flatMap((listItem) => {
+		return listItem.flatMap((componentValue) => {
+			if (
+				!isSimpleBlockNode(componentValue) ||
+				componentValue.startToken[0] !== TokenType.OpenParen
+			) {
+				return [];
+			}
+
+			const blockTokens = componentValue.tokens();
+
+			const mediaQueryList = parseFromTokens(blockTokens, {
+				preserveInvalidMediaQueries: true,
+			});
+
+			return mediaQueryList.filter((mediaQuery) => {
+				return !isMediaQueryInvalid(mediaQuery);
+			});
+		});
+	});
+
+	mediaQueryConditions.forEach((mediaQuery) => {
+		mediaQuery.walk(({ node }) => {
+			if (isMediaFeature(node)) {
+				const token = node.getNameToken();
+
+				if (token[0] !== TokenType.Ident) return;
+
+				callback(token);
+			}
+
+			if (isGeneralEnclosed(node)) {
+				const relevantTokens = topLevelTokenNodes(node);
+
+				if (!relevantTokens) {
+					return;
+				}
+
+				relevantTokens.forEach((token, i) => {
+					if (token[0] !== TokenType.Ident) {
+						return;
+					}
+
+					const nextToken = relevantTokens[i + 1];
+					const prevToken = relevantTokens[i - 1];
+
+					if (
+						// Media Feature
+						(!prevToken && nextToken && nextToken[0] === TokenType.Colon) ||
+						// Range Feature
+						(nextToken &&
+							nextToken[0] === TokenType.Delim &&
+							rangeFeatureOperator.test(nextToken[4].value)) ||
+						// Range Feature
+						(prevToken &&
+							prevToken[0] === TokenType.Delim &&
+							rangeFeatureOperator.test(prevToken[4].value))
+					) {
+						callback(token);
+					}
+				});
+			}
+		});
+	});
+
+	return {
+		stringify() {
```

Comment:
[ask] What reason is to return a `stringify()` object instead of just a string?

Reply:
Serializing takes time/resources.
So for rules that do not have an autofix option it is a bit wasteful to always serialize.

I don't have a strong opinion on this :)
We can simplify the code a bit and have everything run a bit slower, or more complexity and a bit faster.

Reply:
> So for rules that do not have an autofix option it is a bit wasteful to always serialize.

Make sense. Let's keep it. 👍🏼 

Additionally, leaving a code comment describing a reason for returning the `stringify()` object might be better.